lib/Optimizer/CodeGen/ConvertToQIR.cpp

```diff
@@ -1077,6 +1077,164 @@ class MyToMz : public OpConversionPattern<quake::MyOp> {
+      auto idx =
+          rewriter.create<LLVM::ConstantOp>(loc, rewriter.getI64Type(), i);
+      auto call = rewriter.create<LLVM::CallOp>(
+          loc, i8PtrTy, getSymbolRef, ArrayRef<Value>{controlsArr, idx});
+      Value pointer = rewriter.create<LLVM::BitcastOp>(
+          loc, cudaq::opt::factory::getPointerType(i8PtrTy), call.getResult());
+      auto cast = rewriter.create<LLVM::BitcastOp>(loc, i8PtrTy, control);
+      rewriter.create<LLVM::StoreOp>(loc, cast, pointer);
+      i++;
+    }
+
+    return createQubitsCall.getResult();
+  }
+
+  LogicalResult
+  matchAndRewrite(quake::UnitaryOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto ctx = op.getContext();
+    auto parentModule = op->getParentOfType<ModuleOp>();
+    auto constantUnitary = op.getConstantUnitary();
+    auto opName = op.getOpName();
+
+    auto qirArrayTy = cudaq::opt::getArrayType(ctx);
+    auto i8PtrTy =
+        cudaq::opt::factory::getPointerType(IntegerType::get(ctx, 8));
+    FlatSymbolRefAttr symbolRef = cudaq::opt::factory::createLLVMFunctionSymbol(
+        cudaq::opt::QIRArrayCreateArray, qirArrayTy,
+        {rewriter.getIntegerType(32), rewriter.getIntegerType(64)},
+        parentModule);
+    FlatSymbolRefAttr getSymbolRef =
+        cudaq::opt::factory::createLLVMFunctionSymbol(
+            cudaq::opt::QIRArrayGetElementPtr1d, i8PtrTy,
+            {qirArrayTy, rewriter.getIntegerType(64)}, parentModule);
+
+    if (!constantUnitary.has_value()) {
+      auto unitary = adaptor.getUnitary();
+      if (!unitary)
+        return op.emitOpError("no unitary data provided.");
+
+      auto targetsArr = getQubitArray(loc, rewriter, adaptor.getTargets(),
+                                      symbolRef, getSymbolRef);
+      auto controlsArr = getQubitArray(loc, rewriter, adaptor.getControls(),
+                                       symbolRef, getSymbolRef);
+
+      auto unitarySymbolRef = cudaq::opt::factory::createLLVMFunctionSymbol(
+          cudaq::opt::QIRUnitary, LLVM::LLVMVoidType::get(ctx),
+          {unitary.getType(), qirArrayTy, qirArrayTy}, parentModule);
+
+      SmallVector<Value> funcArgs{unitary, controlsArr, targetsArr};
+      rewriter.replaceOpWithNewOp<LLVM::CallOp>(op, TypeRange{},
+                                                unitarySymbolRef, funcArgs);
+
+      return success();
+    }
+
+    auto unitary = constantUnitary.value();
+    // Create the global array of data
+    // Fill the arrays with the unitary data
+    SmallVector<double> realPart(unitary.size()), imagPart(unitary.size());
+    for (std::size_t i = 0; auto &element : unitary) {
+      auto values = dyn_cast<DenseF32ArrayAttr>(element).asArrayRef();
+      realPart[i] = values[0];
+      imagPart[i] = values[1];
+      i++;
+    }
+
+    std::string opNameReal = llvm::formatv("{0}_real", opName);
+    std::string opNameImag = llvm::formatv("{0}_imag", opName);
```

Comment:
```suggestion
    std::string opNameReal = opName.str() + "_real";
    std::string opNameImag = opName.str() + "_imag";
```
Don't think we need the overhead of fancy formatters to concat these.